Fix warning when ball joint has an axis

marcoag opened the pull request:
Closes https://github.com/RobotLocomotion/drake/issues/18951.

Ball joints should have no axis and if there is one the sdf parser will ignore it.

This PR corrects the warning the user gets when the sdf parser finds an axis in a ball joint. It also adds fixes to the relevant tests and removes irrelevant code related to parameters of a ball joint axis.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/18966)
<!-- Reviewable:end -->

EricCousineau-TRI reviewed (commented):
+@EricCousineau-TRI <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> feature, thanks!
+@rpoyner-tri for platform review, please!

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NQ6lBXw1pvcngepKObz:b-cyb6sm)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @rpoyner-tri)



<!-- Sent from Reviewable.io -->

marcoag commented:
Sorry for the confusion here but it seems that axis are still needed for ball joints. They are indeed ignored but some of the parameters specified through them are still relevant, i.e. [damping](https://github.com/RobotLocomotion/drake/blob/b3aea181f2a7bb9775d0aa42b3250ecea792d762/multibody/parsing/detail_sdf_parser.cc#L410). 

I removed all the previous changes and added a warning that clarifies the axis will be ignored but the parameters will be used. I also clarified a bit the comment that explain axis on ball joint.

Because this seems to be a confusing thing from the SDFormat library I opened an issue [here](https://github.com/gazebosim/sdformat/issues/1250) in order to discuss some further clarifications over there.

EricCousineau-TRI reviewed (commented):
Sweet, thanks! I have a question about when this warning will get issues
@rpoyner-tri Can I ask if my question makes sense (and if you have any inklings on the answer)?

> Reviewed 1 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NV6oTKh558bu4cXdDaA:b9u8c7j)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 413 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6o83W-j7eMba0g5Ga:-NV6o83W-j7eMba0g5Gb:bv8upbn) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/detail_sdf_parser.cc#L413)):*
> ```cpp
>   // it should not have actuation, nor should there be a second axis.
>   if (joint_spec.Type() == sdf::JointType::BALL) {
>     if (joint_spec.Axis(0) != nullptr) {
> ```

I'm not sure if I understand when this warning will trigger, and if it will be superfluous.

If a user specifies damping / dynamic parameters, but does not specify `<xyz>`, does the warning get triggered?
If so, is there any way we could ignore it?

e.g., only emit a warning if the axis value itself is explicitly specified? If we don't have access to that, can we at least suppress the warning if it's a default value?
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 1153 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6nry6-LNhWOiQ2Ghq:-NV6nry6-LNhWOiQ2Ghr:b-5pqy05) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/test/detail_sdf_parser_test.cc#L1153)):*
> ```cpp
>       " parameters and limits will be considered.*"));
>   FlushDiagnostics();
> ```

BTW Extra newline?


<!-- Sent from Reviewable.io -->

rpoyner-tri reviewed (commented):
@EricCousineau-TRI I'm not sure exactly what you are asking. The unit test constructs cases where the warnings are issued. Like any other parser warnings, they don't cause the parse to fail, and could be ignored if you need to ignore them. If the concern is console spam, I don't have a good sense of how many ball joints with damping are out there in the world.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NVBD-S39t4VBYlNpXZb:b92mz9b)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)



<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I took Eric's comment to ask whether or not this SDFormat would have any warnings.  I didn't immediately find any test case on point for it:

```xml
  <joint name="jerry" type="ball">
    <parent>world</parent>
    <child>larry</child>
    <axis>
      <limit>
        <effort>0</effort>
      </limit>
    </axis>
  </joint>
```

Note that it's unactuated, with no xyz values given for any axes.  I think what we want is no warnings in this case.

rpoyner-tri reviewed (commented):
Ok then. So that looks like a unit test that needs to get written: given the model shown above, demonstrate no ball axis warnings are issued.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NVBQug415vd-51Ea3G3:bz1g3ul)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)



<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
Sorry for not being clearer, but exactly! Thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NVBZpMU9pY4MAqqwY03:b-mhifpv)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 413 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6o83W-j7eMba0g5Ga:-NVBZtGmDVYdZnUGmEi2:bwro6ga) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/detail_sdf_parser.cc#L413)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I'm not sure if I understand when this warning will trigger, and if it will be superfluous.

If a user specifies damping / dynamic parameters, but does not specify `<xyz>`, does the warning get triggered?
If so, is there any way we could ignore it?

e.g., only emit a warning if the axis value itself is explicitly specified? If we don't have access to that, can we at least suppress the warning if it's a default value?
</blockquote></details>

Macro - would you be able to add in the test that Jeremy mentioned?


<!-- Sent from Reviewable.io -->

marcoag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 413 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6o83W-j7eMba0g5Ga:-NVdAEjSAGONALuVil2u:b7blb24) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/detail_sdf_parser.cc#L413)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Macro - would you be able to add in the test that Jeremy mentioned?
</blockquote></details>

The thing is that according to the SDFormat specification`<xyz>` is required for any `<joint>` `<axis>`: http://sdformat.org/spec?ver=1.9&elem=joint#axis_xyz. So technically you would not be able to specify damping /dynamic parameters without an`<xyz>`. Anytime you have an `<axis>` the warning would be there.
___
*[`multibody/parsing/test/detail_sdf_parser_test.cc` line 1153 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6nry6-LNhWOiQ2Ghq:-NVEUdpFC2Vcwn8K76ED:bss4a2b) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/test/detail_sdf_parser_test.cc#L1153)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

BTW Extra newline?
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966)* status: 1 unresolved discussion, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 413 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6o83W-j7eMba0g5Ga:-NVduWYq1cUdyjmE5ydr:b-z6z3gq) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/detail_sdf_parser.cc#L413)):*
<details><summary><i>Previously, marcoag (Marco A. Gutierrez) wrote…</i></summary><blockquote>

The thing is that according to the SDFormat specification`<xyz>` is required for any `<joint>` `<axis>`: http://sdformat.org/spec?ver=1.9&elem=joint#axis_xyz. So technically you would not be able to specify damping /dynamic parameters without an`<xyz>`. Anytime you have an `<axis>` the warning would be there.
</blockquote></details>

For the record, I agree it's a moot point and the code is OK as-is.


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
> Reviewed 1 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @marcoag)

___
*[`multibody/parsing/detail_sdf_parser.cc` line 413 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-NV6o83W-j7eMba0g5Ga:-NVjPsoFE4IwuygVLa0_:b-r1ktqm) ([raw file](https://github.com/RobotLocomotion/drake/blob/973d14b8e3f97a6110150f240851f2ac3f52e5cd/multibody/parsing/detail_sdf_parser.cc#L413)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

For the record, I agree it's a moot point and the code is OK as-is.
</blockquote></details>

OK Ah, that makes sense - thanks for clarifying!


<!-- Sent from Reviewable.io -->

rpoyner-tri reviewed (commented):
+(status: squashing now) 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18966#-:-NWDJhPN-a6yn8CWqChc:bm9o57a)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),EricCousineau-TRI(platform)



<!-- Sent from Reviewable.io -->
